### PR TITLE
Img upload refactor 2 JEL-86

### PR DESCRIPTION
**Pull request by alexkrolick**

### CODE REVIEW

File: app/javascript/components/ImageModal.js [hunk 1]
```diff
@@ -1,46 +1,55 @@
...
+      setIsAddingImage(true)
+    }
+    window.addEventListener('ImageUploadRequested', handleEvent)
+    return () => window.removeEventListener('ImageUploadRequested', handleEvent)
+  }, [])
 
   const token = document.head.querySelector('[name~=csrf-token][content]')
     .content
 
-  var uppy = Uppy({
```
Comment (alexkrolick) [inline app/javascript/components/ImageModal.js:15]: memoized this so it doesn't create new Uppys on each render

Comment (dennyluan) [reply app/javascript/components/ImageModal.js:15]: nice!

**Comment by dennyluan** [comment]
It looks like the useMemo() here isn't working and breaks on more than 1 uploads from the modal.

**Comment by dennyluan** [comment]
This seems to work 

```javascript
// imagemodal.js
  const success = () => setIsAddingImage(false)

  const uppy = React.useMemo(() => {
    const _uppy = Uppy({
      meta: { type: 'figure' },
      restrictions: { maxNumberOfFiles: 1 },
      autoProceed: true,
    }).use(XHRUpload, {
      endpoint: '/images/store',
      bundle: true,
      headers: {
        csrf: token,
      },
    })

    _uppy.on('complete', (result) => {
      const fileUrl = result.successful[0].response.body.url

      window.dispatchEvent(
        new CustomEvent('ImageUploadCompleted', { detail: { fileUrl, success: () => success() }})
      )
    })


```


```javascript

export const addFigure = function (state, dispatch) {
  if (!dispatch) return true

  window.dispatchEvent(new CustomEvent('ImageUploadRequested'))

  const handleUpload = (evt) => {
    console.log("hi am handling")
    const imageUrl = evt.detail.fileUrl
    const imageType = schema.nodes.image
    evt.detail.success()
    dispatch(state.tr.replaceSelectionWith(imageType.create({ src: imageUrl })))
  }
  window.addEventListener('ImageUploadCompleted', handleUpload, { once: true })

  return true
}
```

**Comment by cindywu** [comment]
this PR will be reviewed june 23rd

**Pull request closed without merging**